File: c++/Sorting/src/mergesort.cpp
#include "mergesort.hpp"
#include <iostream>

MergeSort::MergeSort() {

}

MergeSort::~MergeSort() {

}

int* MergeSort::sort(int array[], int low, int high) {
    if(low < high) {
        int middle = (low+high)/2;
        sort(array,low,middle);
        sort(array,middle+1,high);
        merge(array, low, middle, high);
    }
    return array;
}

void MergeSort::merge(int array[], int low, int middle, int high) {
    int i = low, j = middle+1, k = low;
    int temp[high+1];
    while (i<= middle && j <= high) {
        if(array[i] < array[j]){
            temp[k++] = array[i++];
        } else {
            temp[k++] = array[j++];
        }
    }
    while (i <= middle) {
        temp[k++] = array[i++];
    }
    while (j <= high) {
        temp[k++] = array[j++];
    }
    for(int i = low; i<= high; i++) {
        array[i] = temp[i];
    }
}

// void MergeSort::merge(int array[], int low, int middle, int high) {
//     int size1 = middle - low + 1;
//     int size2 = high - middle;

//     /* Create temp arrays */
//     int left[size1];
//     int right[size2];
//     /* Copy data to temp arrays */
//     for (int i=0; i<size1; ++i)
//         left[i] = array[low + i];
//     for (int j=0; j<size2; ++j)
//         right[j] = array[middle + 1+ j];
//     /* Merge the temp arrays */
//     // Initial indexes of first and second subarrays
//     int i = 0, j = 0;
//     // Initial index of merged subarry array
//     int k = low;
//     while (i < size1 && j < size2) {
//         if (left[i] <= right[j]) {
//             array[k] = left[i++];
//         }
//         else  {
//             array[k] = right[j++];
//         }
//         k++;
//     }
//     /* Copy remaining elements of L[] if any */
//     while (i < size1)  {
//         array[k++] = left[i++];
//     }
//     /* Copy remaining elements of R[] if any */
//     while (j < size2) {
//         array[k++] = right[j++];
//     } 
// }

File: python/Algorithms/Sorting/mergesort.py
class MergeSort:
    
    def __init__(self):
        pass
    
    def sort(self, arr):
        self.mergesort(arr, 0, len(arr)-1)
        
        
    def mergesort(self, arr, left, right):
        if(left >= right):
            return
        
        mid = int((left+right)/2)
        
        self.mergesort(arr, left, mid)
        self.mergesort(arr, mid+1, right)
        self.merge(arr,left, right, mid)
        
    def merge(self, arr, left, right, mid):
        i = left
        j = mid+1
        k = left

        temp = [0 for i in range(right+1)]
        
        while(i <= mid and j <= right):
            if(arr[i] < arr[j]):
                temp[k] = arr[i]
                k += 1
                i += 1
            else:
                temp[k] = arr[j]
                k += 1
                j += 1
        
        while(i <= mid):
            temp[k] = arr[i]
            k += 1
            i += 1
            
        while(j <= right):
            temp[k] = arr[j]
            k += 1
            j += 1
            
        for i in range(left, right+1):
            arr[i] = temp[i]

            